peer/test_utils.go

```diff
@@ -757,6 +757,37 @@ func startPeer(t *testing.T, mockConn *mockMessageConn, peer *Brontide) {
 		mockConn.readMessages, b.Bytes(), make(chan struct{}),
 	)
 	require.True(t, ok)
+
+	// We check here if we have  peer_storage feature enabled as well as
+	// this peer's backup data to determine if we are to expect more
+	// messages on startup.
+	if peer.cfg.PeerDataStore == nil {
+		return
+	}
+
+	backupData, err := peer.cfg.PeerDataStore.Retrieve()
+	require.NoError(t, err)
+
+	if backupData == nil {
+		return
+	}
```

Comment:
This implicitly turns all uses of `startPeer` into tests of the PeerStorage feature. I do not think this is a good idea. I don't want the entire Brontide test suite to fail if this feature doesn't work.

Reply:
I really don't think it does though. It checks early on if PeerDataStore is nil and then returns

Reply:
The testing of the peer data store should not be implicit based off of it being nil or not.

Reply:
I don't follow

Reply:
This method tests the functionality of the peer data store, it is named `startPeer`, these two things should not both be true. You said that it doesn't test the the data store functionality if the db reference is nil, which is true, but this is a bad design as it makes the behavior very implicit. Instead you should break out the peer data store testing functionality into its own method and call it when you want to test that functionality rather than rolling it into this method here.

Reply:
I think it is still in line with the name `startPeer`, init message is sent and received on start up and that is included in the function and as a result of the PR a new message is sent on start up as well which we have included now.

Unless you are suggesting something like this?

```go
startPeer(t *testing.T, mockConn *mockMessageConn,
	peer *Brontide, extraChecks ...func() error)
```

then we can pass in a function to test the sending the peerStorageRetrieval message to the function.

Or what do you suggest?

Reply:
Never do the "extraChecks" thing. I've seen you suggest that a couple of times now. It's never a good idea.

>  Instead you should break out the peer data store testing functionality into its own method and call it when you want to test that functionality rather than rolling it into this method here.

This is what I suggested 👆🏼

Reply:
I need to call it it inside the function as explained in the previous comment (I also explained why the approach in this PR is appropriate). A code snippet of what you are referring to would be helpful

Reply:
Indeed we can break it out. 

But how about this?
> I think it is still in line with the name `startPeer`, init message is sent and received on start up and that is included in the function and as a result of the PR a new message is sent on start up as well which we have included now.

We now have a new message that is conditionally sent on startup, just like init message is included in this function, I think we should include the check for that feature as well. What do you think?

Reply:
> We now have a new message that is conditionally sent on startup

See but the issue here is that the init message is *always* sent, regardless of circumstance. This one is not. As you note, it is *conditionally* sent on startup. This difference is significant. This message is not sent under any of the following circumstances:

1. The feature is not supported
2. There is no channel between the nodes
3. No peer backup message has been stored

Do you see the difference? Embedding a test of a conditionally used feature in the main-line startPeer function is a poor software organization choice.

In general, functions should be *small*. They should do *one* thing and not secretly hide other functionality inside them just because they happen to be called in sequence. In this case you are "secretly hiding" the peer backup storage message testing functionality in the startup sequence and in my view you shouldn't. You could just as easily grab this section of the function and put it into a different function like `checkPeerBackupMessage` and call these two functions back to back, *when appropriate*.

I don't expect you to necessarily know that at your current experience level, and that's fine! 😁 However, I'm asking you to process the reasoning I gave above, and take my word for it. If you don't believe me, I encourage you to try your way of software architecture in your own *personal projects*, and I think you'll find over time that I'm right about this.

These sorts of choices are subtle and they take some time to develop a taste for. However they have been independently rediscovered numerous times, perhaps most famously as the [S in SOLID](https://en.wikipedia.org/wiki/Single-responsibility_principle)